pennylane/spin/lattice.py

```diff
@@ -0,0 +1,196 @@
+        self.n_sites = np.prod(L) * self.n_sl
+
+        if isinstance(boundary_condition, bool):
+            boundary_condition = [boundary_condition for _ in range(self.n_dim)]
+
+        self.boundary_condition = boundary_condition
+        self._test_input_accuracy()
+
+        self.lattice_points, lattice_map = self._generate_grid(neighbour_order)
+
+        cutoff = neighbour_order * np.linalg.norm(self.unit_cell, axis=1).max() + distance_tol
+        edges = self._identify_neighbours(cutoff)
+        self._generate_true_edges(edges, lattice_map, neighbour_order)
+
+    def _test_input_accuracy(self):
+        r"""Tests the accuracy of the input provided"""
+
+        for l in self.L:
+            if (not isinstance(l, np.int64)) or l <= 0:
+                raise TypeError("Argument `L` must be a list of positive integers")
+
+        if self.unit_cell.ndim != 2:
+            raise ValueError("'unit_cell' must have ndim==2, as array of primitive vectors.")
+
+        if self.basis.ndim != 2:
+            raise ValueError("'basis' must have ndim==2, as array of initial coordinates.")
+
+        if self.unit_cell.shape[0] != self.unit_cell.shape[1]:
+            raise ValueError("The number of primitive vectors must match their length")
+
+        if not all(isinstance(b, bool) for b in self.boundary_condition):
+            raise ValueError(
+                "Argument 'boundary_condition' must be a bool or a list of bools of same dimensions as the unit_cell"
+            )
+
+        if len(self.boundary_condition) != self.n_dim:
+            raise ValueError(
+                "Argument 'boundary_condition' must be a bool or a list of bools of same dimensions as the unit_cell"
+            )
+
+    def _identify_neighbours(self, cutoff):
+        r"""Identifies the connections between lattice points and returns the unique connections
+        based on the neighbour_order"""
+
+        tree = cKDTree(self.lattice_points)
+        indices = tree.query_ball_tree(tree, cutoff)
+        unique_pairs = set()
+        edges = {}
+        for i, neighbours in enumerate(indices):
+            for neighbour in neighbours:
+                if neighbour != i:
+                    pair = (min(i, neighbour), max(i, neighbour))
+                    if pair not in unique_pairs:
+                        unique_pairs.add(pair)
+                        dist = np.linalg.norm(
+                            self.lattice_points[i] - self.lattice_points[neighbour]
+                        )
+                        # Scale the distance
+                        bin_density = 21621600  # multiple of expected denominators
+                        scaled_dist = np.rint(dist * bin_density)
+
+                        if scaled_dist not in edges:
+                            edges[scaled_dist] = []
+                        edges[scaled_dist].append((i, neighbour))
+        return edges
+
+    def _generate_true_edges(self, edges, map, neighbour_order):
```

Comment:
Is this function really needed?

Reply:
identify_neighbours function, identifies the edges and sorts them into a dictionary but it doesn't renumber the wrapped lattice points in case of boundary_conditions being True, and doesn't add that extra coordinate to define neighbour order, so we do that in this function. These two can be combined as well